File: src/main/java/com/vc/hard/L924.scala
package com.vc.hard

object L924 {
  def minMalwareSpread(arr: Array[Array[Int]], initial: Array[Int]): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, mutable.ListBuffer[Int]]()

    arr.indices.foreach(i => {
      arr(i).indices.foreach(j => {
        if(arr(i)(j) == 1) {
          val list = map.getOrElse(i, new mutable.ListBuffer[Int]())
          list += j
          map.put(i, list)
        }
      })
    })

    var maxSpread = -1
    var maxSpreadNode = 0
    var spread = 0
    val visited = new mutable.HashSet[Int]()

    def dfs(from: Int): Unit = {
      if(!visited.contains(from)) {
        spread += 1
        visited += from
        if(map.contains(from)) {
          map(from).foreach(to => {
            dfs(to)
          })
        }
      }
    }

    initial.foreach(i => {
      spread = 0
      dfs(i)
      visited.clear
      if(spread > maxSpread) {
        maxSpreadNode = i
        maxSpread = spread
      }
      else if(spread == maxSpread && maxSpreadNode > i) {
        maxSpreadNode = i
      }
    })
    maxSpreadNode
  }
}

File: src/main/java/com/vc/hard/L924.java
package com.vc.hard;

import java.util.*;

class L924 {
    public int minMalwareSpread(int[][] graph, int[] initial) {
        HashMap<Integer, List<Integer>> map = new HashMap<>();

        for(int i = 0; i < graph.length; i++) {
            for(int j = 0; j < graph[0].length; j++) {
                if(graph[i][j] == 1) {
                    List<Integer> list = map.getOrDefault(i, new ArrayList<>());
                    list.add(j);
                    map.put(i, list);
                }
            }
        }

        int maxSpread = -1, maxSpreadNode = 0;
        for(int n: initial) {
            HashSet<Integer> visited = new HashSet<>();
            int spread = solve(n, visited, map);
            if(maxSpread < spread) {
                maxSpread = spread;
                maxSpreadNode = n;
            }
            else if(maxSpread == spread && maxSpreadNode > n) {
                maxSpreadNode = n;
            }
        }
        return maxSpreadNode;
    }

    private int solve(int node, HashSet<Integer> visited, HashMap<Integer, List<Integer>> graph) {
        if(visited.contains(node)) return 0;

        visited.add(node);
        List<Integer> neighbour = graph.getOrDefault(node, new ArrayList<Integer>());
        int res = 1;
        for(Integer i: neighbour) {
            res += solve(i, visited, graph);
        }
        return res;
    }
}
